python/cuml/preprocessing/TargetEncoder.py

```diff
@@ -0,0 +1,213 @@
+            train[self.fold_col] = cp.arange(len(train)) / \
+                (len(train)/self.n_folds)
+            train[self.fold_col] = train[self.fold_col] % self.n_folds
+        elif self.split == 'interleaved':
+            train[self.fold_col] = cp.arange(len(train))
+            train[self.fold_col] = train[self.fold_col] % self.n_folds
+        else:
+            msg = ("split should be either 'random'"
+                   " or 'continuous' or 'interleaved', "
+                   "got {0}.".format(self.split))
+            raise ValueError(msg)
+
+        train[self.fold_col] = train[self.fold_col].astype('int32')
+        self.mean = train[self.y_col].mean()
+
+        cols = [self.fold_col]+x_cols
+
+        agg_each_fold = train.groupby(cols).agg({self.y_col:
+                                                 ['count',
+                                                  'sum']}).reset_index()
+        agg_each_fold.columns = cols + ['count_y', 'sum_y']
+
+        agg_all = agg_each_fold.groupby(x_cols).agg({'count_y': 'sum',
+                                                     'sum_y': 'sum'})\
+                                               .reset_index()
+        cols = x_cols
+        agg_all.columns = cols + ['count_y_all', 'sum_y_all']
+
+        agg_each_fold = agg_each_fold.merge(agg_all, on=x_cols, how='left')
+        agg_each_fold['count_y_all'] = agg_each_fold['count_y_all'] - \
+            agg_each_fold['count_y']
+        agg_each_fold['sum_y_all'] = agg_each_fold['sum_y_all'] - \
+            agg_each_fold['sum_y']
+        agg_each_fold[self.out_col] = (agg_each_fold['sum_y_all'] +
+                                       self.smooth*self.mean) / \
+                                      (agg_each_fold['count_y_all'] +
+                                       self.smooth)
+        agg_each_fold = agg_each_fold.drop(['count_y_all',
+                                            'count_y',
+                                            'sum_y_all',
+                                            'sum_y'], axis=1)
+
+        agg_all[self.out_col] = (agg_all['sum_y_all'] + self.smooth *
+                                 self.mean) / \
+            (agg_all['count_y_all'] + self.smooth)
+        agg_all = agg_all.drop(['count_y_all', 'sum_y_all'], axis=1)
+        self.agg_all = agg_all
+
+        cols = [self.fold_col]+x_cols
+        train = train.merge(agg_each_fold, on=cols, how='left')
+        del agg_each_fold
+        return self._get_return_value(train), train
+
+    def _check_is_fitted(self):
+        if not self._fitted or self.train is None:
+            msg = ("This LabelEncoder instance is not fitted yet. Call 'fit' "
+                   "with appropriate arguments before using this estimator.")
+            raise NotFittedError(msg)
+
+    def _is_train_df(self, df):
+        if len(df) != len(self.train):
+            return False
+        for col in df.columns:
+            if col not in self.train.columns:
```

Comment:
do we support `DataFrame.equals` in cudf?

Reply:
yes, but the `self.train` has more columns so I did the per column comparison only for original columns.

Reply:
I'm a little worried this sort + check will be a bit expensive relative to the cost of transform... maybe I'm crazy though?